### PR TITLE
Upgrade internal and demo app dependencies (close #172)

### LINKED ISSUE #172: Upgrade internal and demo app dependencies
Upgrade dependencies in the demo apps and target the Android SDK 31.

### PR DESCRIPTION
**Pull request by matus-tomlein**
Addresses issue #172.

1. Upgrades dependencies in the demo apps
2. Targets Android SDK v31 in the demo apps

There are a lot of bumped dependencies. Unfortunately, I wasn't able to isolate the changes from each other as they are quite interlinked.

The Detox CI build still targets Android API 28. I wasn't able to update this as there is an [open issue in Detox](https://github.com/wix/Detox/issues/2612) which causes the CI actions to fail with newer SDKs.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ jobs:
 
   build_ios:
     needs: ["test"]
-    runs-on: macos-10.15
+    runs-on: macos-latest
```
Comment (adatzer) [inline .github/workflows/build.yml:71]: would it make sense to pin to `macos-11` instead of `latest`?
(also applies for `deploy.yml`)

File: .github/workflows/e2e-android.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ on:
 
 jobs:
   e2e:
-    runs-on: macos-10.15
+    runs-on: macos-latest
```
Comment (adatzer) [inline .github/workflows/e2e-android.yml:8]: here too about pinning macos version

File: .github/workflows/e2e-ios.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ on:
 
 jobs:
   e2e:
-    runs-on: macos-10.15
+    runs-on: macos-latest
```
Comment (adatzer) [inline .github/workflows/e2e-ios.yml:8]: here too about pinning macos version

File: DemoApp/android/gradle.properties [hunk 4]
```diff
@@ -10,7 +10,7 @@
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
 # Default value: -Xmx10248m -XX:MaxPermSize=256m
-# org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
Comment (adatzer) [inline DemoApp/android/gradle.properties:13]: out of curiosity, why is this change needed?

**Review by adatzer — COMMENTED** [review]
Looks great!
We could also update Micro to latest v1.3.3!

File: .github/workflows/build.yml (hunk 1 above)
Comment (matus-tomlein) [reply .github/workflows/build.yml:71]: I don't have a strong opinion which one to use, changed it to `macos-11`.

File: DemoApp/android/gradle.properties (hunk 4 above)
Comment (matus-tomlein) [reply DemoApp/android/gradle.properties:13]: It was running out of memory sometimes when running the Detox tests. It didn't happen all the time but it made the tests more flaky, so I thought it'd be better just to increase the JVM limit to 2GB.

**Comment by matus-tomlein** [comment]
Thanks @adatzer for the review and comments! Updated the Micro review as well.

File: .github/workflows/build.yml (hunk 1 above)
Comment (matus-tomlein) [reply .github/workflows/build.yml:71]: Unfortunately, when I changed it to `macos-11`, the Android e2e CI action started to fail. Detox couldn't connect to the Android emulator. The same was happening when I used `macos-12`. Couldn't figure out why, it seems like a house of cards 😢  I had a similar problem with Detox last week, it was doing the same thing when I tried to build with Android SDK 31 or 32. After a couple of days debugging, I found out that it is an open issue in Detox. I think we'll just need to keep it like this and wait for a more stable version of Detox.

**Review by adatzer — APPROVED** [review]
Only nits left:
1. Switching to latest also [in this line](https://github.com/snowplow/snowplow-react-native-tracker/blob/66eb1dfa1f860b371447276fbc2b9c3f68ce8593/.github/workflows/deploy.yml#L110)
2. Micro update also [here](https://github.com/snowplow/snowplow-react-native-tracker/blob/66eb1dfa1f860b371447276fbc2b9c3f68ce8593/DemoApp/bin/docker-run-micro.js#L19)

Looks great!!
